hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/listener/SharedTopicListener.java

```
@@ -40,19 +48,71 @@
 
     @Override
     public Flux<TopicMessage> listen(TopicMessageFilter filter) {
-        UnicastProcessor<String> processor = UnicastProcessor.create();
-        Flux<String> timeoutFlux = processor.delayElements(listenerProperties.getBufferTimeout())
-                .replay(1)
-                .autoConnect();
+        TimeoutContext timeoutContext = new TimeoutContext(Schedulers.parallel(),
+                listenerProperties.getBufferTimeout());
+        Mono<TopicMessage> timeoutMono = Mono.create(timeoutContext);
 
-        return getSharedListener(filter)
-                .publishOn(Schedulers.boundedElastic())
+        Flux<TopicMessage> topicMessageFlux = getSharedListener(filter)
                 .doOnSubscribe(s -> log.info("Subscribing: {}", filter))
-                .doOnCancel(() -> processor.onNext("timeout"))
+                .doOnCancel(timeoutContext::onCancel)
                 .onBackpressureBuffer(listenerProperties.getMaxBufferSize())
-                .timeout(timeoutFlux, message -> timeoutFlux, Mono.error(
-                        new ClientTimeoutException("Client timed out while consuming the buffered messages")));
+                .doOnCancel(timeoutContext::onComplete)
+                .doOnError(t -> timeoutContext.onComplete())
+                .doOnComplete(timeoutContext::onComplete);
+        return Flux.merge(listenerProperties.getPrefetch(), topicMessageFlux, timeoutMono)
```

This is quite a complicated solution for not much gain in functionality. Why do we even care about letting them consume the buffer after they overflow? There's no recovery logic here so we should just fail fast and let the client retry.

---

it's for the case that the topic TPS is high, the client is slow and can't catch up, since the backpressure buffer is big, letting the client consume the remaining data then terminating it with either the OverflowException or the ClientTimeoutException would end in a rate higher than 500 the client can get by retry.

---

Regardless if we have a buffer timeout or not, the client has to retry right? They can't keep up and the TPS is high. All we're doing is delaying it slightly and give them a little bit higher TPS on the first run but then still slow on the retry anyway. I don't think an extra 2s window provides any value.